olp-cpp-sdk-core/src/thread/Continuation.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (C) 2022 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <atomic>
+#include <memory>
+#include <thread>
+#include <utility>
+
+#include <olp/core/thread/Continuation.h>
+
+#include <olp/core/client/CancellationToken.h>
+#include <olp/core/thread/SyncQueue.h>
+#include <olp/core/thread/TaskScheduler.h>
+
+namespace olp {
+namespace thread {
+
+using internal::ContinuationImpl;
+
+namespace internal {
+
+class Processor {
+  using Task = std::function<void()>;
+  struct ProcessorInternal;
+
+ public:
+  Processor(std::shared_ptr<thread::TaskScheduler> task_scheduler,
+            ExecutionContext execution_context,
+            std::deque<ContinuationImpl::ContinuationTask> tasks,
+            ContinuationImpl::FinalCallbackType final_callback)
+      : task_scheduler_(task_scheduler),
+        processor_(std::make_shared<ProcessorInternal>(
+            task_scheduler_, std::move(tasks),
+            std::move(std::move(final_callback)),
+            std::move(execution_context))) {}
+
+  // returns false when the current task is over and the queue should proceed
+  // further, true means do nothing
+  bool ProcessAsyncTask(std::shared_ptr<ProcessorInternal> context_copy,
+                        std::pair<ContinuationImpl::AsyncTaskType,
+                                  ContinuationImpl::TaskType>& task) {
+    auto& current_task = task;
+    if (!current_task.first) {
+      // go to the next task in the queue to finilize the task
+      context_copy->tasks_.pop();
+      return false;
+    }
+
+    auto callback = [this, context_copy, &current_task](void* arg) mutable {
+      // limit lifetime of the context shared_ptr to the function scope, for
+      // extra safety in case user's async function stores callback somewhere
+      // for forever
+      auto context = std::move(context_copy);
```

Comment:
Capturing something in the lambda, even if you think you move it actually you only move a copy of it. Are you sure this actually leaves the lambda captured variable empty?